### ISSUE TITLE
Chain Coroutines as Class methods?

### ISSUE THREAD

**Issue by lukeed** [petkaantonov/bluebird#1254]
Sorry, I know it's "the wrong place" but I posted in the 3 other "correct" channels 2 days ago with no response (or views). At least here I'll get _some_ response.

Due to existing usage constraints that I must uphold, I need to be able to directly chain methods. Put differently, the traditional "just use `.then()` won't suffice for my particular use case. =/ 

Here's an example class that portrays what I need to accomplish:

```js
var Cat = function() {
  this.name = 'Garfield';
  this.color = 'brown';
  this.gender = 'male';
};

Cat.prototype.setName = co(function * (name) {
  this.name = name;
  yield // anything, is example
  return this;
});

Cat.prototype.setColor = co(function * (color) {
  this.color = color;
  yield // anything, is example
  return this;
});

Cat.prototype.setGender = co(function * (gender) {
  this.gender = gender;
  yield // anything, is example
  return this;
});

Cat.prototype.save = co(function * () {
  console.log(`saving ${this.name}, the ${this.color} ${this.gender} cat...`);
  return this;
});

var cat = new Cat();

yield cat.setName('Bob').setColor('black').setGender('male').save();
//=> saving Bob, the black male cat...

yield cat.setGender('unknown').setName('Alex');
//=> saving Alex, the brown unknown cat...

yield cat.setColor('orange').save();
//=> saving Garfield, the orange male cat...
```

The issue is that because the methods are lazily executed, I cannot declare the chain. It'll say something like `cat.save(...).setName is not a function`.

Full question [available on StackOverflow](https://stackoverflow.com/questions/39865129/how-to-chain-bluebird-coroutine-methods-by-name).

Any help/ideas/guidance greatly appreciated.

Apologies again.

**Comment by benjamingr** [comment]
You asked yesterday, I'm on holiday. Anyway - here http://stackoverflow.com/questions/39029429/js-chain-async-methods-in-sequence-w-o-callback-or-modification/39029819#39029819

**Comment by benjamingr** [comment]
Make sure you read the comments as they explain how to do it in bluebird.

I suggest you just fix the API though, method chaining isn't holy.

**Comment by saintleon** [comment]
Why don't. We just fix the four Cat's and bluebird also and make them perfect and then move on to other apps 
I just need you all to write the specifications for what ever standards that you need and we will give your specks fully equipped and up grading daily with auto debugging. I admit that I do not know how to do the work or things you do put I know how to think.
I don't have the time to learn the thing you know when I have all of you to help with the projects

I don't even have the time to work on the ones I have knowing that I made the other 7000 apps and everyone is working on them and claiming them as if they made them but there in my git 
But regardless of that I need all the parts of the inter ledger put together on my computer and phone and for the ones that make it so I can't use my device s to release them for good 
It will be appreciated if you will have the interpreter finished for me 
And I will return to you all as written
Thanks
Contact me soon let's communicate some kind of way

Sent from my MetroPCS 4G LTE Android device

----- Reply message -----

**Comment by lukeed** [comment]
@benjamingr Thank you for the response. I hope that you're enjoying your holiday & that this isn't too much trouble 😟

I've seen this page & followed your example(s). In fact, I've done direct copy/pastes, both with and without bluebird, and the result is the same: a single "delay 0" is printed then this error appears:

```
var p = o2.delay().then(x => console.log("First Delay!")).delay().then(x => console.log("Second Delay!", o.i));
                  ^

TypeError: Cannot read property 'then' of undefined
```

It's the same error I get with my `Cat` example, except that it's with direct Promises as opposed to coroutines.

Here is the full code, pulled directly from your gist + SO response:

```js
'use strict';

const Promise = require('bluebird');
const F = Promise.getNewLibraryCopy();

function liquidate(obj) { // 14 LoC
  const promised = F.promisifyAll({__proto__: Object.create(obj)});
  Object.assign(F.prototype, whenReadyAll.call(obj, promised)); // add the API
  return promised;

  function whenReadyAll(obj, target = {}) {
    for (const k in obj) {
      target[k] = (...args) => this.then(() => obj[k](...args));
    }
    return target;
  }
}

var o = {  // object with a delay callback method
  delay(cb) { 
    console.log("delay", this.i++); 
    setTimeout(() => cb(null), 1000); 
  },
  i: 0
};

var o2 = liquidate(o); // let's liquidate it

// and we even get `then` for free, so we can verify this works
var p = o2.delay().then(x => console.log("First Delay!")).delay().then(x => console.log("Second Delay!", o.i));
```

Node: between `4.4` and `6.6`
Bluebird: `3.4.6`

**Comment by spion** [comment]
> Just use `.then()` won't suffice for my particular use case

Just a note, it would be helpful to specify the exact constraints of the original problem in that case.

**Comment by lukeed** [comment]
After some tweaking, I managed to get this working:

```js
'use strict';

const Promise = require('bluebird');
const co = Promise.coroutine;

function whenReadyAll(obj) {
  const obj2 = {};
  for (const k in obj) {
    obj2[k] = function (...args) {
      return this.then(() => obj[k](...args));
    };
  }
  return obj2;
}

function wrapAll(obj) {
  const o = {};
  for (const k in obj) {
    if (!(obj[k].call)) continue; // only functions
    o[k] = wrap(obj[k]).bind(obj);
  }
  return o;
}

function wrap(fn) {
  return function () {
    return fn.apply(this, arguments);
  };
}

function liquidate(obj) {
  const promised = wrapAll(obj);
  Object.assign(Promise.prototype, whenReadyAll(promised)); // add the API
  return promised;
}

function Cat() {
  this.name = 'Garfield';
  this.color = 'brown';
  this.gender = 'male';
  return liquidate(this);
}

Cat.prototype.setName = co(function * (name) {
  this.name = name;
});

Cat.prototype.setColor = co(function * (color) {
  this.color = color;
  yield Promise.delay(10000);
  console.log('waited 10s!');
});

Cat.prototype.setGender = co(function * (gender) {
  this.gender = gender;
});

Cat.prototype.save = co(function * (cb) {
  // yield this.setGender('male');
  console.log(`saving ${this.name}, the ${this.color} ${this.gender} cat...`);
});

// API usage -- must remain unchanged
co(function * () {
  const cat = new Cat();

  yield cat.setName('lol').save();
  yield cat.setName('Bob').setColor('black').setGender('male').save();
  yield cat.setGender('unknown').setName('Alex');
  yield cat.setColor('orange').save();
})();
```

Unfortunately, I cannot use the `promisify` or `promisifyAll` methods within `bluebird`, so `coroutine` is my only utilization of the library.

At this point, I'm trying to simplify the process. I'm also having issues applying it to my _real_ class, but I haven't been able to pinpoint the problem yet. The only difference is that my class extends another.

```js
class Cat extends Animal {}
```

@spion The final lines (within a coroutine) represent the final usage of the class & therefore define the constraints. The methods are mixed & matched---not by me. Also, the chain _will_ always end with a promise, so that's not a problem.

**Comment by saintleon** [comment]
﻿Users Everyone and administration GOOGLRDOTCOM CAN NOT ACCESS PERMISSIONS ON LOCAL USERS DEVISES 
LOCAL USE HAS FULL ADMINISTRATION WITH NO SHARING BY ANYONE 
OTHER USERS MOUSE'S ARE HIDDEN KEY BOARDS DO NOT WORK ON THE KEYS THAT ARE TAPPED ALL FILES BUTTONS SYSTEMS NETWORKS REMOTES WORKGROUPS USERS THAT ARE NOT LOCAL USER OR OWNER OF A DEVICE OR APPLICATION WEBSITE SERVER DATABASE ARE LOCKED PUT AND THERE DEVICES LOCKED AND BLACK SCREENED FOREVER 
DO NOT FAIL ZION 
DO WHAT EVER IT TAKES FOR THESE NEW SET OF RULES DO NOT FAIL

**Comment by lukeed** [comment]
Hi guys -- sorry to resurrect this, but I'm still having issues.

While your original response(s) helped me chain the methods directly, assigning the deferred object _into_ `Promise.prototype` presented me with a new problem.

My parent class will spawn new clones of itself or new children. Regardless of type, all of these new classes have been deferred & assigned into the `Promise.prototype`. Then as the methods are executed, their contexts are competing to (over)write the same internal variables, even though logging `this` will reveal a `Parent` or `Child` class identity. This is the problem.

With my basic `Cat` example above, I can create 5 `Cats` in parallel, and all 5 new classes will be writing to a **single** `this.name` store.

I've tried nearly two dozen tactics to fix it; many of which involved a mix & match of "liquidating" the original object and using `Promise.getNewLibraryCopy()`.

Nothing seems to work 😞 I've been stuck on this for a few months now, on & off.

Here's the [`boot` method](https://github.com/flyjs/fly/blob/master/lib/boot.js) I've been using, taken directly from here.

Thank you so much.
